codejail/jail.py

```diff
@@ -0,0 +1,329 @@
+                elif os.path.isfile(filename):
+                    shutil.copy(filename, homedir)
+                else:
+                    shutil.copytree(filename, dest, symlinks=True)
+
+            # Create the main file.
+            if code:
+                with open(os.path.join(homedir, "jailed_code"), "wb") as jailed:
+                    jailed.write(code)
+
+                argv = ["jailed_code"] + argv
+
+            # Create extra files requested by the caller:
+            for name, content in extra_files or ():
+                with open(os.path.join(homedir, name), "wb") as extra:
+                    extra.write(content)
+
+            cmd = []
+
+            # Build the command to run.
+            if self.user:
+                # Run as the specified user
+                cmd.extend(['sudo', '-u', self.user, 'TMPDIR=tmp'])
+
+            # Add the command-specific pieces
+            cmd.extend(self.cmdline_start)
+            # Add the code-specific command line pieces.
+            cmd.extend(argv)
+
+            # Use the configuration and maybe an environment variable to determine
+            # whether to use a proxy process.
+            use_proxy = limits.LIMITS["PROXY"]
+            if use_proxy is None:
+                use_proxy = int(os.environ.get("CODEJAIL_PROXY", "0"))
+            if use_proxy:
+                run_subprocess_fn = run_subprocess_through_proxy
+            else:
+                run_subprocess_fn = run_subprocess
+
+            # Run the subprocess.
+            status, stdout, stderr = run_subprocess_fn(
+                cmd=cmd, cwd=homedir, env=env, slug=slug,
+                stdin=stdin,
+                realtime=limits.LIMITS["REALTIME"], rlimits=limits.create_rlimits(),
+            )
+
+            # TODO: limit too much stdout/stderr data?
+            result = JailResult(status, stdout, stderr)
+
+            # Remove the tmptmp directory as the sandbox user since the sandbox
+            # user may have written files that the application user can't
+            # delete.
+            run_subprocess_fn(self._build_rm_command(tmptmp), cwd=homedir)
+
+        return result
+
+    def _build_rm_command(self, directory):
+        """
+        Collect the command line arguments to remove the specified directory
+        """
+        rm_cmd = []
+        if self.user:
+            rm_cmd.extend(['sudo', '-u', self.user])
+        rm_cmd.extend([
+            '/usr/bin/find', directory,
+            '-mindepth', '1', '-maxdepth', '1',
+            '-exec', 'rm', '-rf', '{}', ';'
+        ])
+        return rm_cmd
+
+
+# Configure the commands
+
+# COMMANDS is a map from an abstract command name to a list of command-line
+# pieces, such as subprocess.Popen wants.
+COMMANDS = {}
+
+
+def configure(command, bin_path, user=None, lang=None):
+    """
+    Configure a Jail object
+
+    `command` is the abstract command you're configuring, such as "python" or
```

Comment:
These argument names (other than lang) were pre-existing.  I don't want to break backwards compatibility, except where necessary.